File: quant-data/collector/eastmoney.py
"""
东方财富行情HTTP客户端
纯requests，零外部依赖，无Cookie/Token/签名
所有接口均为 secid 驱动的统一设计
"""
import json
import time
import requests
from typing import Optional


BASE_HIS = "http://push2his.eastmoney.com"
BASE_PUSH = "https://push2.eastmoney.com"

# 必需请求头：东方财富从 2026 年起要求 Referer，否则直接关闭连接
DEFAULT_HEADERS = {
    "Referer": "https://quote.eastmoney.com/",
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36",
}

# secid 前缀映射
def _secid(code: str, tp: str = "stock") -> str:
    """构建东方财富 secid
    - stock: 1.600519(沪) / 0.000001(深/北)
    - index: 1.000001(沪指) / 0.399001(深指)
    - concept: 90.BK0612
    """
    if tp == "concept":
        return f"90.{code}"
    if tp == "index":
        prefix = "1" if code.startswith("000") or code.startswith("688") else "0"
        return f"{prefix}.{code}"
    # stock
    prefix = "1" if code.startswith("6") else "0"
    return f"{prefix}.{code}"


class EastmoneyClient:
    """东方财富行情采集"""

    def __init__(self, retry: int = 3, delay: float = 0.3):
        self.retry = retry
        self.delay = delay  # 请求间隔，防止IP封禁

    def _get(self, url: str, params: dict, retry: int = None) -> dict:
        """带重试和速率控制的GET请求"""
        if retry is None:
            retry = self.retry
        # 请求前等待，控制速率
        time.sleep(self.delay)
        for i in range(retry):
            try:
                resp = requests.get(url, params=params, headers=DEFAULT_HEADERS, timeout=15)
                if resp.status_code == 200:
                    return resp.json()
                # 非200状态码，等待后重试
                time.sleep(1.0 * (i + 1))
            except Exception:
                if i < retry - 1:
                    time.sleep(2.0 * (i + 1))  # 连接失败等更久
        return {}

    # ─── 日K (个股/指数/概念 统一) ───

    def get_daily_kline(self, code: str, tp: str = "stock",
                        days: int = 30) -> list[dict]:
        """
        获取日K线，返回近N天数据
        :param code: 股票/指数/概念代码
        :param tp: stock / index / concept
        :param days: 返回最近多少天
        :return: [{trade_date, open, high, low, close, pre_close,
                   change_pct, volume, amount, turnover}, ...]
        """
        sid = _secid(code, tp)
        # 指数/概念用不同的ut
        ut = "fa5fd1943c7b386f172d6893dbfba10b" if tp in ("index", "concept") \
            else "7eea3edcaed734bea9cbfc24409ed989"
        
        params = {
            "secid": sid,
            "klt": "101",  # 日K
            "fqt": "1",    # 前复权
            "beg": "19900101",
            "end": "20500101",
            "lmt": str(days + 10),
            "fields1": "f1,f2,f3,f4,f5,f6",
            "fields2": "f51,f52,f53,f54,f55,f56,f57,f58,f59,f60,f61,f116",
            "ut": ut,
        }
        data = self._get(f"{BASE_HIS}/api/qt/stock/kline/get", params)
        if not data or not data.get("data"):
            return []
        
        klines = data["data"].get("klines", [])
        result = []
        for line in klines:
            parts = str(line).split(",")
            if len(parts) < 11:
                continue
            result.append({
                "trade_date": parts[0],
                "open": float(parts[1]),
                "close": float(parts[2]),
                "high": float(parts[3]),
                "low": float(parts[4]),
                "volume": int(float(parts[5])),
                "amount": float(parts[6]),
                "change_pct": float(parts[8]) if parts[8] else 0.0,
                "change": float(parts[9]) if parts[9] else 0.0,
                "turnover": float(parts[10]) if len(parts) > 10 and parts[10] else 0.0,
            })
        # 计算 pre_close
        for r in result:
            r["pre_close"] = round(r["close"] - r.get("change", 0), 2)
        
        # 只保留最近N天
        return result[-days:] if len(result) > days else result

    # ─── 分钟K / 当日分时 ───

    def get_minute_trends(self, code: str, tp: str = "stock") -> list[dict]:
        """
        获取当日分时数据 (1分钟粒度)
        :return: [{trade_time, price, open, high, low, volume, amount,
                   avg_price, change, change_pct, pre_close}, ...]
        """
        sid = _secid(code, tp)
        ut = "fa5fd1943c7b386f172d6893dbfba10b" if tp in ("index", "concept") \
            else "fa5fd1943c7b386f172d6893dbfba10b"

        params = {
            "secid": sid,
            "ndays": "1",
            "iscr": "1" if tp == "stock" else "0",
            "fields1": "f1,f2,f3,f4,f5,f6,f7,f8,f9,f10,f11,f12,f13",
            "fields2": "f51,f52,f53,f54,f55,f56,f57,f58",
            "ut": ut,
        }
        data = self._get(f"{BASE_PUSH}/api/qt/stock/trends2/get", params)
        if not data or not data.get("data"):
            return []

        trends = data["data"].get("trends", [])
        pre_close = float(data["data"].get("preClose", 0))
        result = []
        for line in trends:
            parts = str(line).split(",")
            if len(parts) < 8:
                continue
            price = float(parts[2])
            result.append({
                "trade_time": parts[0],
                "open": float(parts[1]),
                "price": price,
                "high": float(parts[3]),
                "low": float(parts[4]),
                "volume": int(float(parts[5])),
                "amount": float(parts[6]),
                "avg_price": float(parts[7]) if parts[7] else price,
                "change": round(price - pre_close, 2),
                "change_pct": round((price - pre_close) / pre_close * 100, 2) if pre_close else 0,
                "pre_close": pre_close,
            })
        return result

    # ─── 概念列表 ───

    def get_concept_list(self) -> list[dict]:
        """
        获取东方财富全部概念板块
        :return: [{code, name, stock_count}, ...]
        """
        all_data = []
        page = 1
        while page < 50:
            params = {
                "pn": str(page),
                "pz": "100",
                "po": "1",
                "np": "1",
                "fid": "f62",
                "fs": "m:90+t:3",
                "fields": "f12,f13,f14,f62",
            }
            data = self._get(f"{BASE_PUSH}/api/qt/clist/get", params, retry=2)
            diff = data.get("data", {}).get("diff", [])
            if not diff:
                break
            for item in diff:
                all_data.append({
                    "code": item.get("f12", ""),
                    "name": item.get("f14", ""),
                    "stock_count": item.get("f62", 0),
                })
            if len(diff) < 100:
                break
            page += 1
        return all_data

    # ─── 概念成分股 ───

    def get_concept_constituents(self, concept_code: str) -> list[dict]:
        """
        获取概念板块成分股
        :return: [{stock_code, stock_name}, ...]
        """
        all_data = []
        page = 1
        while page < 100:
            params = {
                "pn": str(page),
                "pz": "200",
                "fid": "f62",
                "po": "1",
                "np": "1",
                "fltt": "2",
                "invt": "2",
                "fs": f"b:{concept_code}",
                "fields": "f12,f14",
            }
            data = self._get(f"{BASE_PUSH}/api/qt/clist/get", params, retry=2)
            diff = data.get("data", {}).get("diff", [])
            if not diff:
                break
            for item in diff:
                all_data.append({
                    "stock_code": item.get("f12", ""),
                    "stock_name": item.get("f14", ""),
                })
            if len(diff) < 200:
                break
            page += 1
        return all_data

    # ─── 股票列表 (全A股) ───

    def get_stock_list(self) -> list[dict]:
        """
        获取全部A股列表
        fs: m:0+t:6(深主板) + m:0+t:13(创业板) + m:1+t:2(沪主板)
            + m:1+t:23(科创板) + m:0+t:81(北交所)
        """
        markets = [
            ("m:0+t:6", "sz"),     # 深证主板
            ("m:0+t:13", "sz"),    # 创业板
            ("m:1+t:2", "sh"),     # 上证主板
            ("m:1+t:23", "sh"),    # 科创板
            ("m:0+t:81", "bj"),    # 北交所
        ]
        all_stocks = {}
        for fs, exchange in markets:
            page = 1
            while page < 100:
                params = {
                    "pn": str(page),
                    "pz": "500",
                    "po": "1",
                    "np": "1",
                    "fid": "f3",
                    "fs": fs,
                    "fields": "f12,f14",
                }
                data = self._get(f"{BASE_PUSH}/api/qt/clist/get", params, retry=2)
                diff = data.get("data", {}).get("diff", [])
                if not diff:
                    break
                for item in diff:
                    code = item.get("f12", "")
                    if code not in all_stocks:
                        all_stocks[code] = {
                            "code": code,
                            "name": item.get("f14", ""),
                            "exchange": exchange,
                        }
                if len(diff) < 500:
                    break
                page += 1
        return list(all_stocks.values())


if __name__ == "__main__":
    client = EastmoneyClient()
    # 测试个股日K
    daily = client.get_daily_kline("600519", "stock", 5)
    print(f"茅台日K: {len(daily)} 条")
    if daily:
        print(json.dumps(daily[-1], ensure_ascii=False, indent=2))

    # 测试分时
    trends = client.get_minute_trends("600519")
    print(f"\n茅台分时: {len(trends)} 条")
    
    # 测试概念列表
    concepts = client.get_concept_list()
    print(f"\n概念板块: {len(concepts)} 个")

File: astock/internal/fetch/eastmoney.go
package fetch

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/huijiecai/stock/astock/internal/model"
)

type EastMoney struct {
	client  *http.Client
	baseURL string
}

func NewEastMoney() *EastMoney {
	return &EastMoney{
		client:  &http.Client{Timeout: 30 * time.Second},
		baseURL: "https://push2.eastmoney.com/api/qt/stock",
	}
}

func (e *EastMoney) doGet(ctx context.Context, urlStr string) ([]byte, error) {
	req, err := http.NewRequestWithContext(ctx, "GET", urlStr, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "Mozilla/5.0")
	req.Header.Set("Referer", "https://quote.eastmoney.com/")
	resp, err := e.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("eastmoney get: %w", err)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read body: %w", err)
	}
	return body, nil
}

func (e *EastMoney) toSecID(code string, tp model.DataType) string {
	switch tp {
	case model.TypeIndex:
		if strings.HasPrefix(code, "000") || strings.HasPrefix(code, "880") {
			return "1." + code
		}
		return "0." + code
	default:
		return "0." + code
	}
}

func (e *EastMoney) DailyKline(ctx context.Context, code string, tp model.DataType, opts ...Option) ([]model.Bar, error) {
	secID := e.toSecID(code, tp)
	options := &FetchOptions{}
	for _, o := range opts {
		o(options)
	}
	params := url.Values{}
	params.Set("secid", secID)
	params.Set("fields1", "f1,f2,f3,f4,f5,f6")
	params.Set("fields2", "f51,f52,f53,f54,f55,f56,f57,f58,f59,f60,f61")
	params.Set("klt", "101")
	params.Set("fqt", "1")
	if options.Limit > 0 {
		params.Set("lmt", strconv.Itoa(options.Limit))
	} else {
		params.Set("lmt", "30")
	}
	if options.Start != "" {
		params.Set("beg", options.Start)
	}
	if options.End != "" {
		params.Set("end", options.End)
	}
	urlStr := "https://push2his.eastmoney.com/api/qt/stock/kline/get?" + params.Encode()
	body, err := e.doGet(ctx, urlStr)
	if err != nil {
		return nil, err
	}
	var result struct {
		Data struct {
			Klined string `json:"klined"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse kline: %w", err)
	}
	if result.Data.Klined == "" {
		return nil, fmt.Errorf("empty kline data for %s", code)
	}
	lines := strings.Split(strings.TrimSpace(result.Data.Klined), ";")
	bars := make([]model.Bar, 0, len(lines))
	for _, line := range lines {
		parts := strings.Split(line, ",")
		if len(parts) < 11 {
			continue
		}
		bar := model.Bar{
			Code:      code,
			Type:      tp,
			TradeDate: parts[0],
		}
		bar.Open, _ = strconv.ParseFloat(parts[1], 64)
		bar.Close, _ = strconv.ParseFloat(parts[2], 64)
		bar.High, _ = strconv.ParseFloat(parts[3], 64)
		bar.Low, _ = strconv.ParseFloat(parts[4], 64)
		bar.PreClose, _ = strconv.ParseFloat(parts[5], 64)
		bar.ChangePct, _ = strconv.ParseFloat(parts[6], 64)
		bar.Volume, _ = strconv.ParseInt(parts[7], 10, 64)
		bar.Amount, _ = strconv.ParseFloat(parts[8], 64)
		bar.Turnover, _ = strconv.ParseFloat(parts[9], 64)
		bars = append(bars, bar)
	}
	return bars, nil
}

func (e *EastMoney) StockList(ctx context.Context) ([]model.Stock, error) {
	urlStr := "https://push2.eastmoney.com/api/qt/clist/get?" + url.Values{
		"pn": {"1"}, "pz": {"10000"}, "po": {"0"}, "np": {"1"},
		"ut": {"bd1d9ddb04089700cf9c27f6f7426281"},
		"fltt": {"2"}, "invt": {"2"}, "fid": {"f3"},
		"fs": {"m:0+t:6,m:0+t:80,m:1+t:2,m:1+t:23,m:0+t:81+s:2048"},
		"fields": {"f12,f14"},
	}.Encode()
	body, err := e.doGet(ctx, urlStr)
	if err != nil {
		return nil, err
	}
	var result struct {
		Data struct {
			Total int `json:"total"`
			Diff  []struct {
				F12 string `json:"f12"`
				F14 string `json:"f14"`
			} `json:"diff"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse stock list: %w", err)
	}
	stocks := make([]model.Stock, 0, result.Data.Total)
	for _, d := range result.Data.Diff {
		exchange := "sz"
		if strings.HasPrefix(d.F12, "6") || strings.HasPrefix(d.F12, "9") {
			exchange = "sh"
		} else if strings.HasPrefix(d.F12, "8") {
			exchange = "bj"
		}
		stocks = append(stocks, model.Stock{Code: d.F12, Name: d.F14, Exchange: exchange})
	}
	return stocks, nil
}

func (e *EastMoney) ConceptList(ctx context.Context) ([]model.Concept, error) {
	urlStr := "https://push2.eastmoney.com/api/qt/clist/get?" + url.Values{
		"pn": {"1"}, "pz": {"500"}, "po": {"0"}, "np": {"1"},
		"ut": {"bd1d9ddb04089700cf9c27f6f7426281"},
		"fltt": {"2"}, "invt": {"2"}, "fid": {"f3"},
		"fs": {"m:0+t:10"},
		"fields": {"f12,f14,f20"},
	}.Encode()
	body, err := e.doGet(ctx, urlStr)
	if err != nil {
		return nil, err
	}
	var result struct {
		Data struct {
			Total int `json:"total"`
			Diff  []struct {
				F12 string `json:"f12"`
				F14 string `json:"f14"`
				F20 int    `json:"f20"`
			} `json:"diff"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse concept list: %w", err)
	}
	concepts := make([]model.Concept, 0, result.Data.Total)
	for _, d := range result.Data.Diff {
		concepts = append(concepts, model.Concept{Code: d.F12, Name: d.F14, StockCount: d.F20})
	}
	return concepts, nil
}

func (e *EastMoney) ConceptConstituents(ctx context.Context, code string) ([]string, error) {
	urlStr := "https://push2.eastmoney.com/api/qt/clist/get?" + url.Values{
		"pn": {"1"}, "pz": {"1000"}, "po": {"0"}, "np": {"1"},
		"ut": {"bd1d9ddb04089700cf9c27f6f7426281"},
		"fltt": {"2"}, "invt": {"2"}, "fid": {"f3"},
		"fs": {"b:" + code},
		"fields": {"f12"},
	}.Encode()
	body, err := e.doGet(ctx, urlStr)
	if err != nil {
		return nil, err
	}
	var result struct {
		Data struct {
			Diff []struct {
				F12 string `json:"f12"`
			} `json:"diff"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse constituents: %w", err)
	}
	codes := make([]string, 0, len(result.Data.Diff))
	for _, d := range result.Data.Diff {
		codes = append(codes, d.F12)
	}
	return codes, nil
}

func (e *EastMoney) TodayMinute(ctx context.Context, code string, tp model.DataType) ([]model.Tick, error) {
	secID := e.toSecID(code, tp)
	params := url.Values{}
	params.Set("secid", secID)
	params.Set("fields1", "f1,f2,f3,f4,f5,f6,f7")
	params.Set("fields2", "f51,f52,f53,f54,f55")
	params.Set("lmt", "500")
	params.Set("is_cr", "0")
	urlStr := e.baseURL + "/kline/get?" + params.Encode()
	body, err := e.doGet(ctx, urlStr)
	if err != nil {
		return nil, err
	}
	var result struct {
		Data *struct {
			Klined string `json:"klined"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse today minute: %w", err)
	}
	if result.Data == nil || result.Data.Klined == "" {
		return nil, fmt.Errorf("empty today minute data for %s", code)
	}
	lines := strings.Split(strings.TrimSpace(result.Data.Klined), ";")
	ticks := make([]model.Tick, 0, len(lines))
	for _, line := range lines {
		parts := strings.Split(line, ",")
		if len(parts) < 5 {
			continue
		}
		tick := model.Tick{
			Code: code, Time: parts[0],
			Price:    mustFloat(parts[1]),
			AvgPrice: mustFloat(parts[3]),
			Volume:   mustInt(parts[4]),
			Amount:   mustFloat(parts[5]),
		}
		ticks = append(ticks, tick)
	}
	return ticks, nil
}

func (e *EastMoney) RealTimeQuote(ctx context.Context, codes ...string) ([]model.Quote, error) {
	if len(codes) == 0 {
		return nil, nil
	}
	secIDs := make([]string, len(codes))
	for i, c := range codes {
		secIDs[i] = "0." + c
	}
	params := url.Values{}
	params.Set("secid", strings.Join(secIDs, ","))
	params.Set("fields", "f2,f3,f4,f5,f6,f12,f14,f15,f16,f17,f18,f20,f21")
	urlStr := e.baseURL + "/get?" + params.Encode()
	body, err := e.doGet(ctx, urlStr)
	if err != nil {
		return nil, err
	}
	var result struct {
		Data struct {
			Total int `json:"total"`
			Diff  []struct {
				F12 string  `json:"f12"`
				F14 string  `json:"f14"`
				F2  float64 `json:"f2"`
				F3  float64 `json:"f3"`
				F4  float64 `json:"f4"`
				F5  float64 `json:"f5"`
				F6  float64 `json:"f6"`
				F15 float64 `json:"f15"`
				F16 float64 `json:"f16"`
				F17 float64 `json:"f17"`
				F18 float64 `json:"f18"`
				F20 float64 `json:"f20"`
				F21 float64 `json:"f21"`
			} `json:"diff"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse quote: %w", err)
	}
	quotes := make([]model.Quote, 0, len(result.Data.Diff))
	for _, d := range result.Data.Diff {
		quotes = append(quotes, model.Quote{
			Code: d.F12, Name: d.F14, Open: d.F15, High: d.F16, Low: d.F17,
			Price: d.F2, PreClose: d.F18, ChangePct: d.F3,
			Volume: int64(d.F4), Amount: d.F5, HighLimit: d.F20, LowLimit: d.F21,
		})
	}
	return quotes, nil
}

func (e *EastMoney) RankVolume(ctx context.Context, top int) ([]model.Quote, error) {
	limit := top
	if limit <= 0 {
		limit = 30
	}
	params := url.Values{}
	params.Set("pn", "1")
	params.Set("pz", strconv.Itoa(limit))
	params.Set("po", "1")
	params.Set("np", "1")
	params.Set("ut", "bd1d9ddb04089700cf9c27f6f7426281")
	params.Set("fltt", "2")
	params.Set("invt", "2")
	params.Set("fid", "f6")
	params.Set("fs", "m:0+t:6,m:0+t:80,m:1+t:2,m:1+t:23")
	params.Set("fields", "f2,f3,f4,f5,f6,f12,f14,f15,f16,f17,f18,f20,f21")
	urlStr := "https://push2.eastmoney.com/api/qt/clist/get?" + params.Encode()
	body, err := e.doGet(ctx, urlStr)
	if err != nil {
		return nil, err
	}
	var result struct {
		Data struct {
			Diff []struct {
				F12 string  `json:"f12"`
				F14 string  `json:"f14"`
				F2  float64 `json:"f2"`
				F3  float64 `json:"f3"`
				F6  float64 `json:"f6"`
				F15 float64 `json:"f15"`
				F16 float64 `json:"f16"`
				F17 float64 `json:"f17"`
				F18 float64 `json:"f18"`
			} `json:"diff"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse rank volume: %w", err)
	}
	quotes := make([]model.Quote, 0, len(result.Data.Diff))
	for _, d := range result.Data.Diff {
		quotes = append(quotes, model.Quote{
			Code: d.F12, Name: d.F14, Price: d.F2, ChangePct: d.F3,
			Amount: d.F6, Open: d.F15, High: d.F16, Low: d.F17, PreClose: d.F18,
		})
	}
	return quotes, nil
}

func (e *EastMoney) RankLimitUp(ctx context.Context) ([]model.Quote, error) {
	params := url.Values{}
	params.Set("pn", "1")
	params.Set("pz", "200")
	params.Set("po", "0")
	params.Set("np", "1")
	params.Set("ut", "bd1d9ddb04089700cf9c27f6f7426281")
	params.Set("fltt", "2")
	params.Set("invt", "2")
	params.Set("fid", "f3")
	params.Set("fs", "m:0+t:6,m:0+t:80,m:1+t:2,m:1+t:23")
	params.Set("fields", "f2,f3,f4,f12,f14,f15,f16,f17,f18,f20")
	urlStr := "https://push2.eastmoney.com/api/qt/clist/get?" + params.Encode()
	body, err := e.doGet(ctx, urlStr)
	if err != nil {
		return nil, err
	}
	var result struct {
		Data struct {
			Diff []struct {
				F12 string  `json:"f12"`
				F14 string  `json:"f14"`
				F2  float64 `json:"f2"`
				F3  float64 `json:"f3"`
				F4  float64 `json:"f4"`
				F15 float64 `json:"f15"`
				F16 float64 `json:"f16"`
				F17 float64 `json:"f17"`
				F18 float64 `json:"f18"`
				F20 float64 `json:"f20"`
			} `json:"diff"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("parse limit up: %w", err)
	}
	quotes := make([]model.Quote, 0, len(result.Data.Diff))
	for _, d := range result.Data.Diff {
		if d.F3 < 9.5 {
			continue
		}
		quotes = append(quotes, model.Quote{
			Code: d.F12, Name: d.F14, Price: d.F2, ChangePct: d.F3,
			Volume: int64(d.F4), Open: d.F15, High: d.F16, Low: d.F17,
			PreClose: d.F18, HighLimit: d.F20,
		})
	}
	return quotes, nil
}

// compile-time interface check
var _ Fetcher = (*EastMoney)(nil)

// MinuteKline is not supported by EastMoney
func (e *EastMoney) MinuteKline(ctx context.Context, code string, tp model.DataType, freq model.Freq, opts ...Option) ([]model.Bar, error) {
    return nil, fmt.Errorf("EastMoney: MinuteKline not implemented, use TDX")
}

// helpers
func mustFloat(s string) float64 {
	v, _ := strconv.ParseFloat(s, 64)
	return v
}
func mustInt(s string) int64 {
	v, _ := strconv.ParseInt(s, 10, 64)
	return v
}
